src/api/cow-subgraph/index.ts

```diff
@@ -1,76 +1,88 @@
 import log from 'loglevel'
-import fetch from 'cross-fetch'
-import { GraphQLClient, Variables } from 'graphql-request'
+import { Variables, request } from 'graphql-request'
 import { DocumentNode } from 'graphql'
 import { CowError } from '../../utils/common'
-import { Context } from '../../utils/context'
+import { Context, Env } from '../../utils/context'
 import { SupportedChainId as ChainId } from '../../constants/chains'
 import { LastDaysVolumeQuery, LastHoursVolumeQuery, TotalsQuery } from './graphql'
 import { LAST_DAYS_VOLUME_QUERY, LAST_HOURS_VOLUME_QUERY, TOTALS_QUERY } from './queries'
 
-export const subgraphUrls: Record<ChainId, string> = {
-  [ChainId.MAINNET]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow',
-  [ChainId.RINKEBY]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow-rinkeby',
-  [ChainId.GOERLI]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow-goerli',
-  [ChainId.GNOSIS_CHAIN]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow-gc',
+export function getSubgraphUrl(env: Env): Partial<Record<ChainId, string>> {
+  switch (env) {
+    case 'staging':
+      return {
+        [ChainId.MAINNET]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow-staging',
+        [ChainId.GNOSIS_CHAIN]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow-gc-staging',
+      }
+    case 'prod':
+      return {
+        [ChainId.MAINNET]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow',
+        [ChainId.RINKEBY]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow-rinkeby',
+        [ChainId.GOERLI]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow-goerli',
+        [ChainId.GNOSIS_CHAIN]: 'https://api.thegraph.com/subgraphs/name/cowprotocol/cow-gc',
+      }
+  }
 }
 
 export class CowSubgraphApi {
   context: Context
-  clients: Record<ChainId, GraphQLClient>
 
   API_NAME = 'CoW Protocol Subgraph'
 
   constructor(context: Context) {
     this.context = context
-    this.clients = this.createClients()
   }
 
-  private createClients(): Record<ChainId, GraphQLClient> {
-    return {
-      [ChainId.MAINNET]: new GraphQLClient(subgraphUrls[ChainId.MAINNET], { fetch }),
-      [ChainId.RINKEBY]: new GraphQLClient(subgraphUrls[ChainId.RINKEBY], { fetch }),
-      [ChainId.GOERLI]: new GraphQLClient(subgraphUrls[ChainId.GOERLI], { fetch }),
-      [ChainId.GNOSIS_CHAIN]: new GraphQLClient(subgraphUrls[ChainId.GNOSIS_CHAIN], { fetch }),
+  async getBaseUrl(options: SubgraphOptions = {}): Promise<string> {
+    const { chainId: networkId, env = 'prod' } = options
+    const chainId = networkId || (await this.context.chainId)
+    let baseUrl = getSubgraphUrl(env)[chainId]
+    if (!baseUrl) {
+      log.warn(
+        `[subgraph:${this.API_NAME}] No subgraph endpoint for chainId: ${chainId} and environment: ${env}. Switching to production mainnet endpoint`
+      )
+      baseUrl = getSubgraphUrl('prod')[ChainId.MAINNET] || ''
```

Comment:
I rather return null/undefined or throw so caller has to deal with this.
Falling back to a different network is not the behaviour I'd like as the SDK consumer.

When one option is not provided, using the default is expected, but when explicitly set the caller should be informed.

Reply:
yes, makes sense. This will be a breaking change in the next release when handling this

Reply:
Since we are still on RC, I don't think it matters.
At least in my view, while the version is not final, breaking changes should be expected